gallery/tutorials/preprocess_imgs_sim.py

```diff
@@ -82,26 +82,26 @@
 imgs_od = source.images[0].asnumpy()
 
 logger.info("Perform phase flip to input images.")
-source.phase_flip()
+source = source.phase_flip()
 imgs_pf = source.images[0].asnumpy()
 
 max_resolution = 15
 logger.info(f"Downsample resolution to {max_resolution} X {max_resolution}")
 if max_resolution < source.L:
-    source.downsample(max_resolution)
+    source = source.downsample(max_resolution)
 imgs_ds = source.images[0].asnumpy()
 
 logger.info("Normalize images to background noise.")
-source.normalize_background()
+source = source.normalize_background()
```

Comment:
Reviewing the gallery in more detail, do you think this (and the following three preprocessing steps) might be a good place to actually demonstrate the new api is not mutating? Historically I think this particular `downsampling` was an area that raised weird behavior to attention.  Or maybe it should be a second section in the notebook? Like show steps as independent sources, then as sequence of sources, and maybe finally as chained (below).  I would sneak `.cache` after the initial source to make it faster if we go down this road.

It would just require changing the `source =` to `source_nb = ` etc.  The `[0].asnumpy()` can be pushed into the plotting which is already messy. The `[0]` will be redundant with `idm` below anyway.

We can show the ability to chain.

`chained_source = source.phase_flip().normalize_background().whiten(noise_estimator).invert_contrast()`

What do you think?

Reply:
Talked with Vineet, I'll pick this gallery edit up as a follow up task.